very early preview: refactor send logic towards PSBT based signing

guggero opened the pull request:
Things don't work or even compile yet, just putting this up to let y'all know what changes are coming and what the state of the refactor is.

dstadulis commented:
Comment to motivate review:
This PR will be a good PR to review for any service (new ones, but also LL products like Loop and Pool) that'll hook into taro logic and need to handle taro transaction events. 👀

Roasbeef commented:
With the current state, perhaps this can be split into _two_ PRs? So one that does the refactor, and another that adds the PSBT stuff.

Roasbeef commented on taropsbt/interface.go:
```diff
@@ -0,0 +1,141 @@
+
+	// IsSplit indicates whether the input asset is being split into more
+	// than one piece. This is always true for non-interactive sends.
+	IsSplit bool
```
I think this should also have an `mssmt.Proof` object embedded? 

Depending on the assumptions/use, it can either have the above, or even just a full `proof.AssetSnapshot`. 

If both sides have already verified the inputs, then this isn't needed. However, if both sides are collaborating without that, then provide the last asset proof for the input either lets them just accept that, or use a Universe to further query backwards for verification.

guggero replied:
This is mostly just a convenience flag. This being false means it's a full value interactive send. So I can also extract that into a helper function so we don't need to encode it in the packet. Yeah, maybe that's a good idea.

Roasbeef commented on taropsbt/interface.go:
```diff
@@ -0,0 +1,141 @@
+	// Asset is the full instance of the asset being spent.
+	Asset *asset.Asset
+
+	// IsSplit indicates whether the input asset is being split into more
```
Would think maybe this can be dropped, but still getting further into the diff.

guggero replied:
I agree, will attempt that.

Roasbeef commented on taropsbt/interface.go:
```diff
@@ -139,3 +143,95 @@ func (o *VOutput) SplitLocator(assetID asset.ID) commitment.SplitLocator {
+
+// extractLocatorFromPath extracts the key family and index from the given BIP32
+// derivation path.
+func extractLocatorFromPath(path []uint32) (keychain.KeyFamily, uint32, error) {
```
Could use a simple comment with the path structure `44/account/index`, etc, etc.

guggero replied:
I agree, will add.

Roasbeef commented on taroscript/send.go:
```diff
@@ -0,0 +1,163 @@
+// or partial amount send) it computes a split commitment with the given input
+// and spend information. The input MUST be checked as valid beforehand and the
+// change output is expected to be declared as such (and be at index 0).
+func PrepareOutputAssets(input *taropsbt.VInput,
```
Cool yeh I think it makes sense to further unify them. We had to go back and make a few quick fixes as our understanding of the protocol evolved during the initial implementation.

Roasbeef commented on taroscript/send.go:
```diff
@@ -0,0 +1,163 @@
+	}
+
+	// We should now have exactly zero value left over after splitting.
+	if residualAmount != 0 {
```
👍

Roasbeef commented on taroscript/send.go:
```diff
@@ -0,0 +1,163 @@
+		// We'll now create a new copy of the old asset, swapping out
+		// the script key. We blank out the tweaked key information as
+		// this is now an external asset.
+		outputs[1].Asset = input.Asset.Copy()
```
Why index 1 here? Thought this was the case of a 1-in-1-out sweep so it can just modify the same output?

guggero replied:
Hmm, right. I just assumed that index 0 would always be the change output on the virtual transaction. And if there is no asset change output, we'd just drop it before signing. But it does make sense to only have one output (index 0) in that case. I guess we should implement an itest for the interactive full value send, as that code path is currently never executed outside of the send unit tests.

Roasbeef commented on taroscript/send.go:
```diff
@@ -0,0 +1,163 @@
+		// asset. This Witness still needs a valid signature for the new
+		// asset to be valid.
+		//
+		// TODO(roasbeef): when we fix #121, then this should also be a
```
Looks closed https://github.com/lightninglabs/taro/issues/121

Roasbeef commented on taroscript/send.go:
```diff
@@ -0,0 +1,163 @@
+		// asset committed, even if it's a zero value (tombstone) split
+		// output for the sender.
+		if outputs[idx].IsChange {
+			outputs[idx].Asset = splitCommitment.RootAsset.Copy()
```
Should this also set the NUMs value and zero value, or it's done at a later step?

guggero replied:
If you look at FromAddress in `taropsbt/address.go`, we initialize the change output with zero value and the NUMS point by default. And if there is a remaining change asset, we set it during assignment of the splits in `PrepareOutputAssets`.

Roasbeef commented on taroscript/send.go:
```diff
@@ -152,6 +153,107 @@ func PrepareOutputAssets(input *taropsbt.VInput,
+	}
+
+	// For each input asset leaf, we need to produce a witness. Update the
+	// input of the virtual TX, generate a witness, and attach it to the
```
I think the intention here is to re-sign all the other leaves that live in the same tree as the input?

guggero replied:
Aaah, I see. Yeah, that makes sense. So we'd actually need to loop over all the assets in the input set. Going to re-add a TODO somewhere for this.

Roasbeef commented on taroscript/send.go:
```diff
@@ -152,6 +153,107 @@ func PrepareOutputAssets(input *taropsbt.VInput,
+	// TODO(guggero): I think this is wrong... We shouldn't look at
+	// PrevWitnesses of the single asset we spend but instead have multiple
+	// inputs if we want to spend multiple coins of the same asset?
+	prevWitnessCount := len(newAsset.PrevWitnesses)
```
Yeah so we need to sign the new asset (which also anchors the split), along with the other leaves that occupied that same prev anchor output. 

Here we're instead, re-signing all the old inputs for a merge (iiuc) which isn't correct. 

In addition to just the `Vinput`, we also need the `TaroCommitment` for that, as for now we need to re-sign all the other inputs. FWIW, we only support spending a single input for now, so things are restricted to just what was in that sole tree.

Roasbeef replied:
Ah ok this was lifted from the old code. It's actually partially untested, since we only support a single input. This would be for the "merge" case, where we need to sign more than one taro input. 

Once we lift the merge restriction (ends up needing a diff proof file creation), then this is equivalent to signing multiple Bitcoin inputs when you need to consolidate change to make a spend.

guggero replied:
Yeah, I'll update the TODO to mention the merge case.

Roasbeef commented on taroscript/send.go:
```diff
@@ -152,6 +153,107 @@ func PrepareOutputAssets(input *taropsbt.VInput,
+	// If the transfer contains no asset splits, we only need to validate
+	// the new asset with its witness attached.
+	if !input.IsSplit {
+		if err := verifySpend(nil); err != nil {
```
We should be able to just make a single call to `verifySpend`, passing in either all the split or non of them.

guggero replied:
Cool, will do that :+1:

Roasbeef commented on taroscript/send.go:
```diff
@@ -254,6 +254,105 @@ func SignVirtualTransaction(input *taropsbt.VInput, outputs []*taropsbt.VOutput,
+// CreateOutputCommitments creates the final set of TaroCommitments representing
+// the asset send. The input TaroCommitment must be set.
+func CreateOutputCommitments(inputCommitment *commitment.TaroCommitment,
+	input *taropsbt.VInput,
```
Seems like we can actually _embed_ `inputCommitment` within `input`?

guggero replied:
I intended to do that initially. But then I figured that the commitment might be enormous, depending on how many assets and asset coins might be contained in it. So I opted to _not_ include that in the packet. Since it's not needed during the signing process of the virtual transaction, this should be fine. But it does mean that the entity that creates the anchor TX must be able to fetch the full input commitment from somewhere. So currently only the owner of the spent asset could do it. But I think with universes being present this information could also be fetched from somewhere.

Roasbeef commented on taroscript/send.go:
```diff
@@ -254,6 +254,105 @@ func SignVirtualTransaction(input *taropsbt.VInput, outputs []*taropsbt.VOutput,
+	// Remove the spent Asset from the AssetCommitment of the sender. Fail
+	// if the input AssetCommitment or Asset were not in the input
+	// TaroCommitment.
+	changeTaroCommitment, err := inputCommitment.Copy()
```
Important to mind all the copies in the old code, as these are in place to avoid unwanted mutations.

guggero replied:
I think I did them justice and carefully adapted all of the calls to `Copy()`. But if you spot one I missed, please let me know!

Roasbeef commented on tarofreighter/chain_porter.go:
```diff
@@ -649,54 +609,21 @@ func (p *ChainPorter) stateStep(currentPkg sendPackage) (*sendPackage, error) {
 		//
 		// TODO(jhb): ScriptKey derivation instructions should be
 		// specified in the AssetParcel
+		var err error
```
Already looking much nicer in this diff!

Still getting through, but I think the extra step here would then be moving a lot of the state like `SenderNewInternalKey` into something that better encapsulates all the data needed to progress with the send.

guggero replied:
Yes, that's done in later commits.

Roasbeef commented on tarofreighter/chain_porter.go:
```diff
@@ -705,17 +632,6 @@ func (p *ChainPorter) stateStep(currentPkg sendPackage) (*sendPackage, error) {
-		}
-
-		currentPkg.SendDelta = preparedSpend
-
```
Looks like we can go ahead and just remove some of these states now as well.

guggero replied:
Yep, see one of the later commits (in the new PRs).

Roasbeef commented on tarofreighter/chain_porter.go:
```diff
@@ -779,13 +693,20 @@ func (p *ChainPorter) stateStep(currentPkg sendPackage) (*sendPackage, error) {
 			return nil, err
 		}
 
+		sendPacket.Outputs[0].TaprootInternalKey = schnorr.SerializePubKey(
```
Any reason this can't be done within the function above?

guggero replied:
This was cleaned up in the current PRs.

Roasbeef commented on tarofreighter/chain_porter.go:
```diff
@@ -606,24 +606,6 @@ func (p *ChainPorter) stateStep(currentPkg sendPackage) (*sendPackage, error) {
-		if err != nil {
-			return nil, err
-		}
-
```
Looks like another state that can just be removed.

Roasbeef commented on taropsbt/address.go:
```diff
@@ -0,0 +1,45 @@
+	// collectible, we will need to create a split with un-spendable change.
+	// Since we don't have any inputs selected yet, we'll use the NUMS
+	// script key to avoid deriving a new key for each funding attempt.
+	pkt.Outputs[0] = &VOutput{
```
Should actually be conditional? So then would only set the receiver output.

guggero replied:
No, not in this case. Since we're creating a PSBT for a send to an address, it will be a non-interactive send and therefore always have a split (but maybe a zero value with NUMS point).

Roasbeef commented on tarofreighter/parcel.go:
```diff
@@ -205,122 +200,13 @@ type sendPackage struct {
-	TransferTx *wire.MsgTx
+	// AnchorTx is the BTC level anchor transaction with all its information
+	// as it was used when funding/signing it.
+	AnchorTx *AnchorTransaction
```
Nice, now we're getting closer to having to carry along minimal state here.

Roasbeef commented on tarofreighter/wallet.go:
```diff
@@ -280,3 +448,148 @@ func inputAnchorPkScript(assetInput *AnchoredCommitment) ([]byte, []byte,
+	// Since we're adding the input of the anchor output of our prior asset
+	// later, we need to add this value here, so we don't lose the amount
+	// to fees.
+	pkt.Pkt.UnsignedTx.TxOut[maxOutputIndex].Value += anchorInputValue
```
I think we'll also want to revisit some of this logic here w/ the amounts and fees as well. When I send on the CLI, the posted chain fees look to be 10x or so higher than I'd expect.

guggero replied:
Yes, definitely! Should be a bit easier now with the new code in place.

Roasbeef commented on tarofreighter/parcel.go:
```diff
@@ -17,46 +17,18 @@ import (
-	// SendStatePreparedSplit prepares the splits (if needed) for a
-	// transfer that will create a change output.
-	SendStatePreparedSplit
+	SendStateVirtualCommitmentSelect SendState = iota
```
😎

Roasbeef reviewed (changes requested):
Excellent work! Love how we were able to simplify much of the logic and also the main state machine, while outlining the idealized virtual PSBT flow as well. Did an initial pass through everything to build up my mental model and be able to better answer architectural questions. 

As mentioned offline, I think we could split this into two PRs: one that does the refactoring, and another the starts t use the new logic and hooks up the PSBT aspects. 

I think we can also further wrap the `sendPackage` struct to have shim calls into the PSBT logic which should also serve to simplify things (no longer need to manually assign those fields, as they become nearly all private and are modified with the entrance methods.

guggero commented:
Replaced by https://github.com/lightninglabs/taro/pull/243, https://github.com/lightninglabs/taro/pull/244 and https://github.com/lightninglabs/taro/pull/245.

@Roasbeef I only saw your review now, not sure how I missed it (I suspect GitHub has somehow been swallowing notifications lately). Will incorporate them into the now split PRs.

guggero commented on taropsbt/interface.go:
```diff
@@ -0,0 +1,141 @@
+
+	// AnchorOutputTaprootBip32Derivation is the Taproot BIP32 derivation of
+	// the anchor output's internal key.
+	AnchorOutputTaprootBip32Derivation *psbt.TaprootBip32Derivation
```
The thing is, since we might need to sign for this key, we need the full, 33-byte internal key so we can use it for the `lnd` APIs. And the Taproot version only contains the 32-byte x-only key, which means we'd not be able to derive the internal key in 50% of the times. Unless we only rely on the derivation information? But I think even there the `lnd` PSBT APIs also want the non-Taproot version to be set.
